packages/evm/src/generic.ts

```diff
@@ -0,0 +1,228 @@
+  ContractAbi extends Abi,
+  FunctionName extends ExtractAbiFunctionNames<ContractAbi, 'nonpayable' | 'payable'>,
+> {
+  // TODO: investigate why "extends BaseTransferParams" is not working
+  source: Domainlike;
+  destination: Domainlike;
+  sourceNetworkProvider: Eip1193Provider;
+  sourceAddress: string;
+  resource: string | EvmResource;
+
+  gasLimit: bigint;
+  functionParameters: AbiParametersToPrimitiveTypes<
+    ExtractAbiFunction<ContractAbi, FunctionName>['inputs'],
+    'inputs'
+  >;
+  functionName: FunctionName;
+  destinationContractAbi: ContractAbi;
+  destinationContractAddress: string;
+  maxFee: bigint;
+}
+/**
+ * Creates a cross chain contract
+ * call object
+ * @param {GenericMessageTransferRequest<ContractAbi, FunctionName>} request
+ * @returns {Promise<GenericMessageTransfer<ContractAbi, FunctionName>>}
+ */
+export async function createCrossChainContractCall<
+  ContractAbi extends Abi,
+  FunctionName extends ExtractAbiFunctionNames<ContractAbi, 'nonpayable' | 'payable'>,
+>(
+  request: GenericMessageTransferRequest<ContractAbi, FunctionName>,
+): Promise<GenericMessageTransfer<ContractAbi, FunctionName>> {
+  const config = new Config();
+  await config.init(process.env.SYGMA_ENV);
+  const genericTransfer = new GenericMessageTransfer<ContractAbi, FunctionName>(request, config);
+
+  const isValidTransfer = await genericTransfer.isValidTransfer();
+  if (!isValidTransfer) throw new Error('Invalid transfer.');
+
+  return genericTransfer;
+}
+/**
+ * GenericMessageTransfer contains
+ * functionality that facilitates generic
+ * transfers or contract calls between two
+ * EVM chains (for now)
+ */
+class GenericMessageTransfer<
+  ContractAbi extends Abi,
+  FunctionName extends ExtractAbiFunctionNames<ContractAbi, 'nonpayable' | 'payable'>,
+> extends BaseTransfer {
+  maxFee: bigint;
+  destinationContractAddress: string;
+  gasLimit: bigint;
+  functionName: FunctionName;
+  destinationContractAbi: ContractAbi;
+  functionParameters: AbiParametersToPrimitiveTypes<
+    ExtractAbiFunction<ContractAbi, FunctionName>['inputs'],
+    'inputs'
+  >;
+  /**
+   * Create a GenericMessageTransfer object
+   * @param {GenericMessageTransferRequest<ContractAbi, FunctionName>} params
+   * @param {Config} config
+   */
+  constructor(params: GenericMessageTransferRequest<ContractAbi, FunctionName>, config: Config) {
+    super(params, config);
+    this.destinationContractAddress = params.destinationContractAddress;
+    this.gasLimit = params.gasLimit;
+    this.functionParameters = params.functionParameters;
+    this.functionName = params.functionName;
+    this.destinationContractAbi = params.destinationContractAbi;
+    this.maxFee = params.maxFee;
+  }
+  /**
+   * Checks whether the transfer is valid
+   * given all parameters
+   * @returns {Promise<boolean>}
+   */
+  async isValidTransfer(): Promise<boolean> {
+    // Resource type should always be generic
+    if (
```

Comment:
what if we return a cast to boolean of this condition? like `!!condition`?

Reply:
wdym?

Reply:
like doing this:

```ts
return !!(this.resource.type !== ResourceType.PERMISSIONED_GENERIC &&
      this.resource.type !== ResourceType.PERMISSIONLESS_GENERIC)
```

It's just a nit